torch/csrc/jit/codegen/cuda/lower_index_hoist.cpp

```diff
@@ -0,0 +1,283 @@
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+// Return leaf domains of a given domain.
+std::unordered_set<IterDomain*> getUsedLeafIds(
+    IterDomain* id,
+    TensorDomain* td) {
+  std::deque<IterDomain*> child_ids;
+  child_ids.push_back(id);
+
+  std::unordered_set<IterDomain*> used_leaf_ids;
+
+  while (!child_ids.empty()) {
+    auto child_id = child_ids.front();
+    child_ids.pop_front();
+    auto it = std::find(td->domain().begin(), td->domain().end(), child_id);
+    if (it != td->domain().end()) {
+      used_leaf_ids.insert(child_id);
+    } else {
+      for (auto expr : child_id->uses()) {
+        for (auto out : expr->outputs()) {
+          TORCH_INTERNAL_ASSERT(out->isA<IterDomain>());
+          child_ids.push_back(out->as<IterDomain>());
+        }
+      }
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      !used_leaf_ids.empty(),
+      "No used id found: ",
+      id->toString(),
+      ", ",
+      td->toString());
+
+  return used_leaf_ids;
+}
+
+} // namespace
+
+CommonIndexKey::CommonIndexKey(
+    IterDomain* consumer_indexed_id,
+    TensorDomain* consumer_td,
+    TensorDomain* ref_td,
+    const std::unordered_map<IterDomain*, Val*>& ref_index_map,
+    const std::vector<kir::ForLoop*>& loops) {
+  auto gpu_lower = GpuLower::current();
+
+  concrete_indexed_id_ =
+      gpu_lower->caIndexMap().getConcreteMappedID(consumer_indexed_id);
+
+  const auto consumer_leaf_ids =
+      getUsedLeafIds(consumer_indexed_id, consumer_td);
+
+  // Convert to Parallel concrete IDs to find matching loops.
+  std::unordered_set<IterDomain*> concrete_leaf_ids;
+  for (auto& id : consumer_leaf_ids) {
+    concrete_leaf_ids.insert(
+        gpu_lower->caParallelMap().getConcreteMappedID(id));
+  }
+
+  // Find used loops and their index vals
+  for (const auto i : c10::irange(loops.size())) {
+    auto loop = loops.at(i);
+    auto loop_id =
+        gpu_lower->caParallelMap().getConcreteMappedID(loop->iter_domain());
+    auto it = concrete_leaf_ids.find(loop_id);
+    if (it != concrete_leaf_ids.end()) {
+      // This leaf reference id is used for indexing the consumer id
+      used_loops_.push_back(loop);
+      auto index_it = ref_index_map.find(ref_td->axis(i));
+      TORCH_INTERNAL_ASSERT(
+          index_it != ref_index_map.end(),
+          "Index not found for leaf ID, ",
+          ref_td->axis(i)->toString());
+      loop_index_vals_.push_back(index_it->second);
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      !used_loops_.empty(),
+      "No loop used for indexing found. ",
+      consumer_indexed_id->toString());
+
+  TORCH_INTERNAL_ASSERT(
+      consumer_leaf_ids.size() == used_loops_.size(),
+      "consumer_leaf_ids.size() = ",
+      consumer_leaf_ids.size(),
+      ", used_loops_.size() == ",
+      used_loops_.size(),
+      ", loops.size() == ",
+      loops.size());
+
+  // If the inner-most loop is vectorized, that loop is not
```

Comment:
That's an interesting point. Didn't think about broadcast. Let me think about it a little more carefully

Reply:
I think we don't need to think about broadcast domains here. These are leaf domains, so if they are broadcast, the corresponding root domains must be broadcast too, and since we don't generate any index for broadcast domains, there's nothing "common" with tensors having non-concretized broadcast domains.

Reply:
That said, I had many cases of missing hoisting opportunities during the development of this PR. Let me know if you find fusion patterns that should use hoisted indices but are not translated to do so.